### PR TITLE
more tests for edge cases and fix edge cases

### PR DESCRIPTION
**Pull request by rileyberton**
Found a few edge cases:

1) getting the same value twice from the LFU would segfault
2) there is a case where buckets would get ordered incorrectly

### CODE REVIEW

File: src/utils/mtev_lfu.c [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 struct lfu_cache_entry;
 
-STAILQ_HEAD(frequency_list, lfu_cache_entry);
+TAILQ_HEAD(frequency_list, lfu_cache_entry);
```
Comment (rileyberton) [inline src/utils/mtev_lfu.c]: need bidirectional queue

File: src/utils/mtev_lfu.c [hunk 2]
```diff
@@ -234,21 +231,27 @@ touch_lfu_cache_no_lock(mtev_lfu_t *c, struct lfu_entry *e)
...
-        next_bucket == STAILQ_FIRST(&c->lfu_frequency_list) ||
+        next_bucket == TAILQ_FIRST(&c->lfu_frequency_list) ||
         next_bucket->frequency != bucket_freq + 1) {
       need_new_bucket = mtev_true;
     }
     struct lfu_cache_entry *empty = remove_from_frequency_list_no_lock(c, e);
     if (need_new_bucket) {
       if (empty != NULL) {
+        /* reuse this bucket instead of allocating */
+        empty->frequency = bucket_freq + 1;
```
Comment (rileyberton) [inline src/utils/mtev_lfu.c]: There may be gaps in the bucket list.

`1 -> 3 -> 7`

If we are in `3` and `3` becomes empty.. when it becomes 4 it needs to be inserted *after* `1`, not at the HEAD which would create the wrong order.

Now we check next and prev buckets.. if we have advanced past `next_bucket` then we treat it as the prior, if not, we treat `prev_bucket` as prior.  `prev_bucket` will always be lower in frequency by definition.

File: src/utils/mtev_lfu.c [hunk 3]
```diff
@@ -234,21 +231,27 @@ touch_lfu_cache_no_lock(mtev_lfu_t *c, struct lfu_entry *e)
...
+        if (next_bucket && next_bucket->frequency < empty->frequency) {
+          insert_to_bucket_list(c, empty, next_bucket);
+        } else if (prev_bucket && prev_bucket->frequency < empty->frequency) {
```
Comment (rileyberton) [inline src/utils/mtev_lfu.c]: Probably should just be `else` instead of `else if`...

File: src/utils/mtev_lfu.c [hunk 4]
```diff
@@ -234,22 +231,30 @@ touch_lfu_cache_no_lock(mtev_lfu_t *c, struct lfu_entry *e)
...
+        next_bucket == TAILQ_FIRST(&c->lfu_frequency_list) ||
         next_bucket->frequency != bucket_freq + 1) {
       need_new_bucket = mtev_true;
     }
     struct lfu_cache_entry *empty = remove_from_frequency_list_no_lock(c, e);
     if (need_new_bucket) {
       if (empty != NULL) {
+        /* reuse this bucket instead of allocating */
+        empty->frequency = bucket_freq + 1;
+        if (next_bucket && next_bucket->frequency < empty->frequency) {
```
Comment (aidancully) [inline src/utils/mtev_lfu.c]: this case should be impossible? `c->lfu_frequency_list` is now stored sorted, so `empty`'s old frequency is guaranteed to be less than `next_bucket->frequency`. Further, we've already checked if `next_bucket->frequency` == new frequency (i.e. old frequency + 1), and it's not, because we `need_new_bucket`... I think we're OK w/ replacing all this logic with the body of the `else` block.

Comment (rileyberton) [inline src/utils/mtev_lfu.c]: yes, I suppose either next_bucket is equal to the new frequency or greater.. It can't be less. I will simplify further.
